selectable_pdf_infra/lambda/selectable_pdf/main.py

```diff
@@ -0,0 +1,180 @@
+'''
+process_pdf
+-----------
+
+Add selectable characters to a PDF. Useful to convert a scanned PDF (i.e. pixel 
+characters) into a selectable PDF.
+
+Algorithm:
+1. load the original PDF from S3 (S3 localtion info located in the SQS message)
+2. read all the page and convert them to image. Indeed, the original PDF might already 
+   have selectable characters, or a mix with pixel characters, so we want to avoid 
+   overlay characters on characters
+3. Add the character to the images. the characters are added by words. For each word, 
+   the best fitting fontsize is computed to the character word length equal the pixel 
+   word length
+4. Save the pdf with selectable characters
+
+Required environment variable
+* OUTPUT_BUCKET
+* DDB_DOCUMENTS_TABLE
+'''
+
+# import modules
+# --------------
+# standard modules
+import os
+import boto3
+import json
+import datetime
+
+# custom modules from layers
+from textracttools import load_json_from_s3
+from helpertools import (
+    get_logger,
+    load_pdf_from_s3, 
+    save_pdf_to_s3, 
+    make_pdf_doc_searchable
+)
+
+# typing
+from typing import Dict, Optional
+
+
+# logger
+# ------ 
+#If no LOG_LEVEL env var or wrong LOG_LEVEL env var, fallback 
+# to INFO log level
+logger = get_logger(os.getenv('LOG_LEVEL', default='INFO'))
+
+
+# lambda entrypoint
+# -----------------
+def lambda_handler(event, context):
+    logger.info('event: {}'.format(event))
+    args = parse_args(event)
+    logger.info(f'args: {args}')
+
+    # Get AWS connectors.
+    ddb_ress = boto3.resource('dynamodb')
+    ddb_doc_table = ddb_ress.Table(args['ddb_documents_table'])
+
+    returns = list()
+    for rec in args['records']:
+        document_id = rec['document_id']
```

Comment:
who is generating this document_id? Is it unique? How do we prevent information override in case of collision?

Reply:
There are no check against collisions. The document_id is unique. it's generated with the upload date (down to the micro second) + 8 random characters from UUID1. Would you still set up an anti-collition check? the ID is one of the key in DynamoDB, so it easy to implement...